### PR TITLE
[export] allow Dim(1,2) for export dynamic shapes

### PR DESCRIPTION
**Pull request by pianpwk**
Current dynamic shapes implementation fixes lower range of Dims to be 2 for analysis, but allows 0/1 shapes during runtime. This leads to failures when initializing Dim(1,2). This PR sets the lower bound to 0, and avoids erroring out when conflicting with the generated (2, maxsize) constraint during analysis.

Also resolves a derived dim constraints issue with the following code:
```
class Bar(torch.nn.Module):
    def forward(self, x, y):
        return x + y[1:]

dx = Dim("dx", min=1, max=3)
ep = export(
    Bar(),
    (torch.randn(2, 2), torch.randn(3, 2)),
    dynamic_shapes=({0: dx, 1: None}, {0: dx+1, 1: None})
)
print(ep.range_constraints)
```

In main:
```
{s0: ValueRanges(lower=2, upper=3, is_bool=False), s0 + 1: ValueRanges(lower=3, upper=4, is_bool=False)}
```

This PR:
```
{s0: ValueRanges(lower=1, upper=3, is_bool=False), s0 + 1: ValueRanges(lower=2, upper=4, is_bool=False)}
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by pianpwk** [comment]
Summary of fixes (beyond setting 2 to 0 and avoiding constraint conflicts):

1. Non-strict export was returning the wrong range_constraints, so we now follow how strict mode does it:

Strict mode looks at [input_shape_constraints](https://github.com/pytorch/pytorch/blob/main/torch/export/dynamic_shapes.py#L809-L836), which comes from user-specified constraints [here](https://github.com/pytorch/pytorch/blob/346140486971bdc87838842aaf5a1f100d5eb4e6/torch/_dynamo/eval_frame.py#L1429-L1433)

Non-strict was following [ShapeEnv.var_to_range](https://github.com/pytorch/pytorch/blob/346140486971bdc87838842aaf5a1f100d5eb4e6/torch/_export/non_strict_utils.py#L246) which is set during analysis, which is done with a lower bound of 2

2. Deserialization performs specialization checks on the stored ValueRanges, which can fail if the lower range is < 2. So we overwrite and clamp the ranges at 2 just for analysis (if they don't specialize to 0/1), then restore the original ValueRanges afterwards: [overwrite](https://github.com/pytorch/pytorch/blob/1059493f2b8c2d78665f7d1eed20dafd04d77670/torch/_export/serde/serialize.py#L1502-L1510) [restore](https://github.com/pytorch/pytorch/blob/1059493f2b8c2d78665f7d1eed20dafd04d77670/torch/_export/serde/serialize.py#L1847)

File: torch/_export/non_strict_utils.py [hunk 1]
```diff
@@ -226,26 +232,53 @@ def make_constraints(
...
+    user_tensor_input_names = {
+        spec.arg.name
+        for spec in input_specs
+        if spec.kind == InputKind.USER_INPUT and isinstance(spec.arg, TensorArgument)
+    }
+    tensor_id_to_dim_constraint: Dict[int, Dict[int, Dict[str, int]]] = defaultdict(
+        defaultdict
+    )
+    for const in constraints:
+        spec = const.serializable_spec
```
Comment (avikchaudhuri) [inline torch/_export/non_strict_utils.py:245]: do you really need `serializable_spec` here? isn't `t_id` and `dim` already available?

File: torch/_export/non_strict_utils.py [hunk 2]
```diff
@@ -226,26 +232,53 @@ def make_constraints(
...
+        tensor_constraints = tensor_id_to_dim_constraint[id(flat_args[input_index])]
         for i, d in enumerate(node.meta["val"].shape):
             if isinstance(d, torch.SymInt):
                 # Look up the range constraint for the symbol corresponding to this shape dimension
                 # and store it indexed by the symbolic expression corresponding to it.
                 # NOTE(avik): Use node._expr instead of node.expr for the lookup here because
                 # we want the symbol, not its replacement, which could be an expression. Maybe
                 # there's a better way to do this, e.g., by (re)computing value ranges for expressions?
-                range_constraints[d.node.expr] = shape_env.var_to_range[d.node._expr]
+                constraint = tensor_constraints[i] if tensor_constraints else None
```
Comment (avikchaudhuri) [inline torch/_export/non_strict_utils.py]: I'm not entirely sure what you're doing here. What makes it necessary to look up the specified range instead of the inferred range here? I think it is because the inferred range min is clamped to 2, have you investigated how to avoid that? I continue to be worried that this is not desirable.

@ezyang might need your help here. Is there something we can do (e.g., mark dynamic dims to be size-like) to avoid this clamping. Is the 0/1 specialization flag + default_range to blame here?

File: torch/_export/serde/serialize.py [hunk 3]
```diff
@@ -1496,9 +1496,19 @@ def deserialize(
                 shape_env=self.shape_env,
             )
             self.symbol_name_to_symbol: Dict[str, sympy.Symbol] = {}
-            self.symbol_name_to_range = {} if symbol_name_to_range is None else symbol_name_to_range
             self.signature = self.deserialize_signature(serialized_graph_module.signature)
             self.constants = deserialize_torch_artifact(constants)
+
+            # deserialization does analysis with checks on 0/1, so we create fake range constraints and
```
Comment (avikchaudhuri) [inline torch/_export/serde/serialize.py:1502]: where does it do this analysis, and is it not possible to avoid that instead of tweaking this logic just so?

File: torch/_export/serde/serialize.py [hunk 4]
```diff
@@ -1496,9 +1496,19 @@ def deserialize(
...
+            # restore the original range constraints afterwards
+            self.symbol_name_to_range = {}
+            if symbol_name_to_range:
+                for k, vr in symbol_name_to_range.items():
+                    lower = int(vr.lower)
+                    if vr.upper >= 2:  # no specialization on 0/1
```
Comment (avikchaudhuri) [inline torch/_export/serde/serialize.py:1508]: This check feels...odd. cc @ezyang to take a look.

**Review by avikchaudhuri — COMMENTED** [review]
I'm glad tests are passing but I'm not confident this is the best we can do. Pulling in @ezyang

**Review by avikchaudhuri — APPROVED** [review]
